### PR TITLE
Add option to get label value from an HTTP header

### PR DESCRIPTION
**Pull request by scoof**
This PR allows the client to send the label value as an HTTP header. Doing it this way, combined with #53 will allow this to be used for alerting in Grafana.

### CODE REVIEW

**Comment by simonpasquier** [comment]
:wave: could you rebase?

**Comment by scoof** [comment]
Rebased

File: injectproxy/routes_test.go [hunk 1]
```diff
@@ -421,3 +422,90 @@ func TestQuery(t *testing.T) {
...
+		expBody      []byte
+	}{
+		{
+			// No "namespace" parameter returns an error.
+			expCode: http.StatusBadRequest,
+		},
+		{
+			// Query with only query
+			labelv:       "default",
+			headerv:      "defaulth",
```
Comment (s-urbaniak) [inline injectproxy/routes_test.go:443]: nit: i think for this test case this should be `headerv:      "",`

**Comment by s-urbaniak** [comment]
@scoof thank you for the contribution! i just have one nit regarding a testcase, else lgtm.

File: injectproxy/routes_test.go (hunk 1 above)
Comment (scoof) [reply injectproxy/routes_test.go:443]: You're absolutely correct....

**Review by simonpasquier — COMMENTED** [review]
Do we want to allow passing the label name only by the header? Currently you would have to set the `--label` flag to a non-empty value which is a bit confusing.

**Comment by scoof** [comment]
If we want to do that, then I think we should have a separate flag to set the prometheus label, different from the query param. That would break backwards compatibility. Thoughts?

**Comment by simonpasquier** [comment]
Personally I wouldn't be too worried about backwards compatibility as long as it's reflected in the changelog. I would expect that one should be able to choose between 1) read from query parameters only 2) read from HTTP header only or 3) read from query parameters then HTTP header if empty.

@s-urbaniak @brancz thoughts?

**Comment by s-urbaniak** [comment]
hm, I think you're right, I was somehow assuming that `r.label` is just being used here, but it is all over the place. Yeah I agree, potentially adding another option just for specifying the query parameter sounds good to me.

**Comment by scoof** [comment]
So --label --query-param and --header.
Would that make sense?
Then --label is required and at least one of --query-param and --header

**Comment by scoof** [comment]
Still need to fix docs

**Comment by scoof** [comment]
I haven't forgotten about this, but I'm quite busy with a new baby.

**Comment by horazont** [comment]
Hey @scoof. I took the liberty to take this off your mental load and rebase it into current master in #64. If you find a minute, I’d be glad if you could confirm that you’re attributed appropriately.

**Comment by scoof** [comment]
@horazont looks fine, thanks!

**Pull request closed without merging**